File: example/motor.js
import { EventEmitter } from 'events';

// For debugging
import { Debug } from './debug.js';
const logger = Debug('mock:motor' + '\t');

export class Motor extends EventEmitter {
  constructor({ id, homePos, maxSpeed = 40, maxAccel = 20 }) {
    super();
    this.id = id;
    this.homePos = homePos;
    this.maxSpeed = maxSpeed;
    this.maxAccel = maxAccel;
    this.currentPos = homePos;
    this.goalPos = homePos;
    this.moving = false;
    this.homing = false;
    this.enabled = false;
    logger(`Motor ${this.id} created with homePos ${this.homePos}`);
  }

  get state() {
    return {
      id: this.id,
      homing: this.homing,
      home: this.home,
      enabled: this.enabled,
      error: this.error,
      moving: this.moving,
      currentPos: this.currentPos,
      goalPos: this.goalPos,
    };
  }

  setPosition(position, spd, acc) {
    const speed = spd || this.maxSpeed;
    const acceleration = acc || this.maxAccel;

    logger(
      `Motor ${this.id} setPosition called with position ${position}, speed ${speed}, acceleration ${acceleration}`,
    );
    this.goalPos = position;
    this.moving = true;

    const distance = Math.abs(this.goalPos - this.currentPos);
    const direction = this.goalPos > this.currentPos ? 1 : -1;
    const interval = 100; // Update every 100 ms
    const step = ((speed * interval) / 1000) * direction; // Degrees per interval

    logger(`Motor ${this.id} will be moving in steps of ${step}`);

    const move = () => {
      if (this.moving) {
        // Check if the motor has reached its goal position
        // logger(`Motor ${this.id} difference is ${Math.abs(this.goalPos - this.currentPos)}`);
        if (Math.abs(this.goalPos - this.currentPos) < Math.abs(step)) {
          this.currentPos = this.goalPos;
          this.moving = false;
          logger(`Motor ${this.id} reached goal position ${this.goalPos}`);
          this.emit('moved', this.id);
          this.emit('pulse', this.id, this.currentPos);
          // Maybe we were homing the robot so update homing
          if (this.homing) {
            this.homing = false;
            logger(`Motor ${this.id} finished homing`);
            this.emit('home', this.id);
          }
        }
        // Check if the motor has been disabled ( occurs on manual disable or robot stop )
        else if (!this.enabled) {
          logger(
            `Motor ${this.id} was moving to ${this.goalPos} but was stoped at ${this.currentPos}`,
          );
          this.goalPos = this.currentPos;
          this.moving = false;
          this.emit('moved', this.id);
          this.emit('pulse', this.id, this.currentPos);
        }
        // Keep going :)
        else {
          this.currentPos = this.currentPos + step;
          //   logger(`Motor ${this.id} moving to ${this.currentPos}`);
          this.emit('pulse', this.id, this.currentPos);
          setTimeout(move, interval);
        }
      }
    };

    move();
  }

  goHome() {
    logger(`Motor ${this.id} going home to ${this.homePos}`);
    this.homing = true;
    this.setPosition(this.homePos);
  }

  resetErrors() {
    logger(`Motor ${this.id} resetting errors`);
    this.error = null;
    this.emit('reset');
  }

  enable() {
    logger(`Motor ${this.id} enabled`);
    this.enabled = true;
    this.emit('enabled');
  }

  disable() {
    logger(`Motor ${this.id} disabled`);
    this.enabled = false;
    this.emit('disabled');
  }

  freeze() {
    logger(`Motor ${this.id} freeze called`);
    this.moving = false;
    this.emit('disabled');
  }

  center() {
    logger(`Motor ${this.id} centering`);
    this.setPosition(0);
  }

  zero() {
    logger(`Motor ${this.id} zeroing position`);
    this.currentPos = 0;
    this.goalPos = 0;
    this.emit('reset');
  }

  startHoming() {
    logger(`Motor ${this.id} starting homing`);
    this.emit('homing');
    this.goHome();
  }
}

export default Motor;

File: example_py/motor.py
from pyee import EventEmitter
from debug import Debug
import threading

logger = Debug('mock:motor\t')

class Motor(EventEmitter):
    def __init__(self, id, homePos, limNeg=None, limPos=None, maxSpeed=40, maxAccel=20):
        super().__init__()
        self.id = id
        self.homePos = homePos
        self.limNeg = limNeg
        self.limPos = limPos
        self.maxSpeed = maxSpeed
        self.maxAccel = maxAccel
        self.currentPos = homePos
        self.goalPos = homePos
        self.moving = False
        self.homing = False
        self.enabled = False
        self.error = None
        logger(f'Motor {self.id} created with homePos {self.homePos}')

    @property
    def state(self):
        return {
            'id': self.id,
            'homing': self.homing,
            'home': self.homePos,
            'enabled': self.enabled,
            'error': self.error,
            'moving': self.moving,
            'currentPos': self.currentPos,
            'goalPos': self.goalPos,
        }

    def validate(self, enabled=False, cleared=False, log=''):
        if enabled and not self.enabled:
            message = f'Please enable before {log}'
            logger(message)
            self.error = 'DISABLED'
            self.emit('meta')
            return False
        if cleared and self.error:
            message = f'Please clear error before {log}'
            logger(message)
            self.error = 'CLEAR_ERROR'
            self.emit('meta')
            return False
        return True

    def set_position(self, position, spd=None, acc=None):
        speed = spd or self.maxSpeed
        acceleration = acc or self.maxAccel

        if not self.validate(enabled=True, cleared=True, log='attempting to set position'):
            return

        logger(f'Motor {self.id} setPosition called with position {position}, speed {speed}, acceleration {acceleration}')
        self.goalPos = position
        self.moving = True

        distance = abs(self.goalPos - self.currentPos)
        direction = 1 if self.goalPos > self.currentPos else -1
        interval = 0.1  # Update every 100 ms
        step = (speed * interval) * direction  # Degrees per interval

        logger(f'Motor {self.id} will be moving in steps of {step}')

        def move():
            if self.moving:
                if abs(self.goalPos - self.currentPos) < abs(step):
                    self.currentPos = self.goalPos
                    self.moving = False
                    logger(f'Motor {self.id} reached goal position {self.goalPos}')
                    self.emit('moved', self.id)
                    self.emit('pulse', self.id, self.currentPos)
                    if self.homing:
                        self.homing = False
                        logger(f'Motor {self.id} finished homing')
                        self.emit('home', self.id)
                elif not self.enabled:
                    logger(f'Motor {self.id} was moving to {self.goalPos} but was stopped at {self.currentPos}')
                    self.goalPos = self.currentPos
                    self.moving = False
                    self.homing = False
                    self.emit('moved', self.id)
                    self.emit('pulse', self.id, self.currentPos)
                else:
                    self.currentPos += step
                    self.emit('pulse', self.id, self.currentPos)
                    threading.Timer(interval, move).start()

        move()

    def go_home(self):
        logger(f'Motor {self.id} going home to {self.homePos}')
        if not self.validate(enabled=True, cleared=True, log='attempting to home'):
            return
        self.homing = True
        self.set_position(self.homePos)

    def reset_errors(self):
        logger(f'Motor {self.id} resetting errors')
        self.error = None
        self.emit('reset')

    def enable(self):
        logger(f'Motor {self.id} enabled')
        self.enabled = True
        self.emit('enabled')

    def disable(self):
        logger(f'Motor {self.id} disabled')
        self.enabled = False
        self.emit('disabled')

    def freeze(self):
        logger(f'Motor {self.id} freeze called')
        self.moving = False
        self.emit('freeze')

    def center(self):
        logger(f'Motor {self.id} centering')
        self.set_position(0)

    def zero(self):
        logger(f'Motor {self.id} zeroing position')
        self.currentPos = 0
        self.goalPos = 0
        self.emit('reset')

    def start_homing(self):
        logger(f'Motor {self.id} starting homing')
        self.emit('homing')
        self.go_home()
